### PR TITLE
WIP: remove celeritas

### PR DESCRIPTION
**Pull request by sethrj**
This improves the usability of g4vg at the cost of removing functionality and duplicating some code from celeritas.

### CODE REVIEW

**Comment by JuanGonzalezCaminero** [comment]
Hi, because `FindCUDA` is deprecated (Used by VecCore) the highest CMake version that we can currently use is `3.26`

**Comment by agheata** [comment]
@sethrj This currently cuts the dependency but does not replace the CELER_XXX macros, so the compilation fails. Are you proposing we take it from here and do the rest of the changes, using a library-local copy of these macros? Just to make sure we don't start doing the same thing on the two sides of the pond. What about the testing macros, should we protect them using some #ifdef guard?

**Comment by drbenmorgan** [comment]
Just for reference, my hacked "standalone" version is on this branch: https://github.com/drbenmorgan/g4vg/commits/standalone/

There are some differences in approach - I went for providing internal workarounds/translations of some of the Celeritas functionality, see https://github.com/drbenmorgan/g4vg/blob/e4b2870b833ac3a9f3d1cb39f579fc6796c5aca4/src/detail/TranslationTypes.hh

I'll have a look at the changes here in more detail.

**Comment by drbenmorgan** [comment]
Having had a quick look, I suspect that all the remaining build failures are down to the `CELER_...` macros and other types. As noted above, my branch dealt with this by providing an internal translation:

https://github.com/drbenmorgan/g4vg/blob/e4b2870b833ac3a9f3d1cb39f579fc6796c5aca4/src/detail/TranslationTypes.hh

which I guess is similar to the header in this PR:

https://github.com/celeritas-project/g4vg/pull/12/files#diff-7b2edb38c8fd49855a9f9835ea4cbeb490e3d01d0d5bb1e14a41a02105e85447

That could therefore be expanded to cover the remaining macros/types, or do you have a different plan/ideas here @sethrj?

**Comment by sethrj** [comment]
I think that's about right, glad I didn't do any more since I had forgotten about your patch. I paused because I wasn't sure if we wanted to delete, copy, and/or refactor the various features... looks like you got most of it though! A few things:
- I don't like that we unconditionally remove the assertions since those can check for changes in geant4 behavior etc.
- We can probably get rid of the type-safe OpaqueId entirely since it's not used for anything other than volume IDs
- I'd rather copy-paste (and add namespaces to) the softeq/demangle
- We could find/replace the few uses of `for ( _ : range(..))` with `for ( ; ; )`
- At the end, let's replace all `CELER_` with `G4VG_`

**Comment by drbenmorgan** [comment]
Sounds good @sethrj - agreed that the assertions are good to have!

**Comment by agheata** [comment]
@sethrj why do we need RDC in this PR since the library does not link with any cuda code?

**Comment by sethrj** [comment]
That's a question for @pcanal : I think it's necessary to correctly link downstream libraries against vecgeom cuda? If he confirms it's not necessary, I can remove it.

**Comment by pcanal** [comment]
TLDR: please use the `CudaRdcUtils.cmake` scaffolding.

For better or worse, keeping the CudaRdc fragment is likely the easiest solution.

It might not be needed but using it might be easier/clearer.

Without it, there will be only one `libg4vg` library that would need to make sure to link against the 'middle' library of VecGeom (without [this VecGeom MR](https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/1243)  it is `libvecgeomcuda_static.a`) and then use it directly in executable one would need to add (to each executable) a direct link to the final library (`libvecgeomcuda.so`).   

Technically we could also 'chance' it and link `libg4vg` to the final library and hope and pray that the `nvlink` output in there is ignored.  However this seems to have been working for shared build but have been shown (see Atlas) to fail in static build in the most annoying, unpredictable, hard to trace, expensive to debug and just plain (expletive deleted to protect the guilty and innocent alike) ways you could imagine.

**Comment by agheata** [comment]
> Without it, there will be only one `libg4vg` library that would need to make sure to link against the 'middle' library of VecGeom (without [this VecGeom MR](https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/1243) it is `libvecgeomcuda_static.a`) and then use it directly in executable one would need to add (to each executable) a direct link to the final library (`libvecgeomcuda.so`).

This sounds like a total overkill, given that g4vg itself does/should NOT require any linking against anything in the vecgeom::cuda namespace. g4vg should only link against libvecgeom.a ! So this library is not affected by nvlink at all, it is a pure host library !

**Comment by pcanal** [comment]
> This sounds like a total overkill, given that g4vg itself does/should NOT require any linking against anything in the vecgeom::cuda namespace. g4vg should only link against libvecgeom.a ! So this library is not affected by nvlink at all, it is a pure host library !

Indeed, if this does not need to link against any `libvecgeomcuda`, it does not need any Rdc support.

**Comment by sethrj** [comment]
My understanding is that, if nothing else, the main "vecgeom" library (when cuda is enabled) has missing symbols that can only be filled by libvecgeomcuda. Is that the only limitation?

**Comment by agheata** [comment]
> My understanding is that, if nothing else, the main "vecgeom" library (when cuda is enabled) has missing symbols that can only be filled by libvecgeomcuda. Is that the only limitation?

AFAIK this should not be the case, if it is, we are in trouble and we should fix it with the highest priority ! It is only `libvecgeomcuda` that should depend on `libvecgeom` and not the other way around

**Comment by agheata** [comment]
@sethrj can you try in this PR to link only against `libvecgeom`? Is there a unit test that you can run in standalone mode (w/o/ Celeritas) ?

**Comment by sethrj** [comment]
Will do. Yes there's a standalone "smoke test" that I should be able to try... but I can't get to it today unfortunately.

**Comment by drbenmorgan** [comment]
@sethrj, my updates to bring in `PrintableLV`, `make_gdml_name` and `TypeDemangler` are on my branch:

https://github.com/drbenmorgan/g4vg/tree/remove-celeritas-stringfuncs

That's based on your branch in this PR, so should merge/ff cleanly.

Let me know if there are any changes needed!

**Comment by sethrj** [comment]
@drbenmorgan Did you forget a commit? I see the new file and its inclusion, but no changes to the rest of the code.

**Comment by drbenmorgan** [comment]
Hmm, looking at the branch triggered a rate limit on GitHub so I don't know if there was/is a glitch on their end, but the tip should be this commit:

https://github.com/drbenmorgan/g4vg/commit/acbf965f877862b8d76c584b3fd2f567e493a0a9

and the most recent three commits have the changes.

**Comment by drbenmorgan** [comment]
@sethrj, trying this out locally before this afternoon I found a couple of compile errors from the use of VecGeom Logger. These were simple to fix, so changes are here in two commits:

https://github.com/drbenmorgan/g4vg/tree/remove-celeritas-logger

Important one is the second, which restricts use to VecGeom 1.2.8 or newer, as that's the version where Logger.h became available.

**Comment by sethrj** [comment]
Thanks @drbenmorgan . I'll pull those in (or you can push to my fork) and will try to get the CI online before our meeting today.

File: external/CMakeLists.txt [hunk 1]
```diff
@@ -37,7 +36,7 @@ g4vg_set_default(CELERITAS_USE_VecGeom ON)
 # Use VecGeom, mm, and double precision
 g4vg_set_default(CELERITAS_CORE_GEO VecGeom)
 g4vg_set_default(CELERITAS_UNITS CLHEP)
-g4vg_set_default(CELERITAS_REAL_TYPE double)
+g4vg_set_default(CELERITAS_double double)
```
Comment (sethrj) [inline external/CMakeLists.txt:39]: Oops need to fix this

File: src/G4VG.hh [hunk 2]
```diff
@@ -38,8 +40,8 @@ struct Options
     //! Perform conversion checks
     bool compare_volumes{false};
 
-    //! TODO: allow client to use a different unit system (default: mm = 1)
-    static constexpr double scale = 1;
+    //! Value of 1mm in native unit system (0.1 for cm)
+    double scale = 1;
```
Comment (sethrj) [inline src/G4VG.hh:44]: Vecgeom "kind of" has a scale flag, but it's only used by VGDML. I think we should either move that flag into the vgdml loader or use it here.

P.S. @agheata @SeverinDiederichs the scale should *probably* be used in the "absolute" (non-relative) bump distances inside vecgeom, since otherwise loading and running the same problem at different length scales gives you different results.
